### PR TITLE
As tibble.cube improvements

### PR DESCRIPTION
**Pull request by gshotwell**
This PR adds support for getting a tibble of the real cube, which is mostly going to be useful for Crunch developers having conversations about what the correct cube numbers are for the various different clients.

### CODE REVIEW

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: 'the cube values is expressed' -> 'the cube values are expressed'

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: We kinda already have this if all of the measures are included in the tibble, no?

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: This is great for MR, but what happens with categorical arrays? Those will have the same name as well even though one dimension is items and the other is categories.

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: We probably _do_ want to include `.unweighted_counts` here, or at least have a way to do so.

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: You should be able to use `as.array(x, measure = m)` here.

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: This is probably scope creep, but after looking through the path for real vs. user cube: the real cube always preserves all missing categories, where as the user one follows what's in `useNA`. I'm not certain that should hold up these improvements (especially since we don't have any evidence that anyone is using this already, nevermind would be confused by the fact that you get different missing behaviors!)

**Review by jonkeane — CHANGES_REQUESTED** [review]
Overall looks good. I do think we're missing some renaming logic similar to what we do with MRs for categorical arrays.

File: R/as-tibble.R
Comment (gshotwell) [reply R/as-tibble.R]: It looks like this doesn't work because of the odd cubes which come off summarize

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: It looks like this is the only reason for this argument. I wonder whether we need this argument at all. Why do you think it should be `FALSE` here?

File: tests/testthat/cubes/cat-x-cat.json
Comment (nealrichardson) [inline tests/testthat/cubes/cat-x-cat.json:1]: These look like the same cube fixtures that are in `crunch/tests/testthat/cubes/`, so maybe those should be moved to `crunch/inst/` or something so they can be shared and not duplicated. The advantage of that would be that we wouldn't forget to update this second copy of the fixtures if/when the cube response API evolves.

File: R/summarize.R
Comment (gshotwell) [reply R/summarize.R]: Mostly it was to not change the summarize behaviour with the updates to `as.tibble.` I think the current behaviour of summarize is that missing values are suppressed and unweighted counts aren't included and I think that that behaviour makes sense because `dplyr::summarize` only returns the summary stats that are included as arguments to summarize. So basically when I did `ds %>% summarize(mean = mean(mpg))` my intuition was that it should return the grouping variables and the `mean` column and not include a bunch of extra metadata.

File: tests/testthat/cubes/cat-x-cat.json
Comment (gshotwell) [reply tests/testthat/cubes/cat-x-cat.json:1]: We decided to put this off, I raised an issue about it in rcrunch.

File: R/as-tibble.R
Comment (nealrichardson) [inline R/as-tibble.R:12]: Should the converse also be true: if the query isn't weighted, don't include .unweighted_counts? I think so; "unweighted" has no meaning when there are no weights, and if you wanted counts, you could ask for them.

File: tests/testthat/cubes/mr-by-cat-profiles-stats.json
Comment (nealrichardson) [inline tests/testthat/cubes/mr-by-cat-profiles-stats.json]: See slack discussion; I think this fixture needs to be sanitized.

File: tests/testthat/test-as-tibble.R
Comment (nealrichardson) [inline tests/testthat/test-as-tibble.R]: Since you only use this `check_cube_match` function in this one place, I'm thinking an explicit/verbose test of "you should get this exact tibble" would be a more transparent way to assert this.

File: tests/testthat/test-as-tibble.R
Comment (nealrichardson) [inline tests/testthat/test-as-tibble.R]: There are three skipped tests/TODOs that this branch should resolve. Perhaps you've already tested the features in other tests. Either way, make it so all the behavior is covered and that there are no more skips before we merge.

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: tests/testthat/cubes/mr-by-cat-profiles-stats.json
Comment (gshotwell) [reply tests/testthat/cubes/mr-by-cat-profiles-stats.json]: I think we can remove it from this repo

File: R/as-tibble.R
Comment (gshotwell) [reply R/as-tibble.R:12]: It seems safer to either include both columns or have an argument which prevents including `.unweighted_counts`. For instance if you relied on `.unweighted_counts` in a script, you wouldn't want that to fail if weighting were removed from the query. Also I like that `as_tibble` gives you everything that's in the cube, and I think we include unweighted counts whether in the cube regardless of whether the query is weighted.

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: Does this function just exist so that it can be documented?

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: use `call.=FALSE` with `stop()`

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: "and the cube values are expressed columns for each measure." ?> "and the cube values are expressed as columns for each measure."

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R:39]: FWIW, there is a `crunch` function `is.selectedDimension` which does exactly this. I'm not sure it's necessary to import and use it when you can accomplish it in one line, but it would be using the approved machinery from `crunch` so if that ever changes for some reason (I can't imagine what that would be!) it would be automatically updated.

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: Am I reading this correctly that you're stringing together the missing elements with `|` between them? Any reason not to keep them as a vector?

File: R/as-tibble.R
Comment (jonkeane) [inline R/as-tibble.R]: The global assignment here gives me a little pause — is it possible to do this without it, or at least a bit more constrained than global assignment?

File: R/summarize.R
Comment (jonkeane) [inline R/summarize.R]: I know that it means formula here, but I cannot stop reading this as the [Family and Medical Leave act](https://en.wikipedia.org/wiki/Family_and_Medical_Leave_Act_of_1993)

**Review by jonkeane — COMMENTED** [review]
Here are a few small comments, I need a bit more time to poke at it myself, though.

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: IMO this section would read more cleanly as:

```
out <- as_tibble(crtabs(fmla, data=.data))
# Comment about map_df
unweighted_n <-  map_df(counts, ~ out$row_count)
out$row_count <- NULL
out <- bind_cols(out, unweighted_n)
```

Avoids the `@` into the cube and looks more explicit to me.

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: I found `counts` and `non_counts` hard to follow in the code below. I had to keep scrolling up to see what they really meant, which is `unweighted_n_measures` and `measures`, or something like that.

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: 1) This is a weird edge case. I'd move the large comment block inside the `if`, and then start the `else` below with `# The usual case`, just to make that clear.

2) Technically you could construct a cube query to get this, but `nrow()` is much more intuitive.

3) Explain here why you're doing map_df, for the pathological case where not only does the user request `unweighted_n()` and nothing else but they also request it multiple times! Then you don't have to explain it below.

File: R/summarize.R
Comment (nealrichardson) [inline R/summarize.R]: Note here that if your dataset is already unweighted, it is equivalent to `n()`

File: tests/testthat/test-as-tibble.R
Comment (nealrichardson) [inline tests/testthat/test-as-tibble.R]: I still would prefer to see more explicit tests here. If I change something and this test fails, I'm going to have a hard time seeing what I broke.

File: tests/testthat/test-summarize.R
Comment (nealrichardson) [inline tests/testthat/test-summarize.R]: Break these into separate `test_that`s with descriptive names.

File: tests/testthat/test-summarize.R
Comment (nealrichardson) [inline tests/testthat/test-summarize.R]: So after the extra gymnastics to make sure that if you ask for unweighted_n more than once, it doesn't break, we should have a test for that :)

**Review by nealrichardson — CHANGES_REQUESTED** [review]
We're close!

File: R/as-tibble.R
Comment (gshotwell) [reply R/as-tibble.R:39]: I don't think it's exported, so I'll just keep this for now.

File: R/as-tibble.R
Comment (gshotwell) [reply R/as-tibble.R]: I might be doing a bit much with the pipeline here. The reduce is reducing between the logical vectors in the data.frame, so what you get out is a single logical vector representing whether there are any TRUEs in the underlying dimension. I think this is a safe way of doing it but let me know if you can think of a case I'm missing here. 

```
x@dims %>% (list)
            map("missing") %>% (list of short logical vectors)
            expand.grid() %>% (data frame each column is a dimension, row is the unique combination of dimensions)
            reduce(`|`) ( df$v1 | df$v2 | dfv3 .... | df$vn)
```

Initially I kept each dimension's missingness as its own column, but in working with the output I found that I only ever wanted to know whether any dimension was missing for a given entry. The thought was that if any dimension is missing then the entry is also missing right?

File: R/as-tibble.R
Comment (gshotwell) [reply R/as-tibble.R]: Is there something you're worried about in particular? The thought was that it's super-assigned to the function's argument so I don't think it can affect state outside of the function. It's possible to do the same thing with a for loop, but the reason I prefer the `walk` pattern is that it makes it clear that you're calling the function for its side effects, and it's also clear which part of the function is impure.  I can't think of a case where the loop would generate a different value from `walk` + `<<-`

File: R/summarize.R
Comment (gshotwell) [reply R/summarize.R]: That's the correct reading.

File: R/summarize.R
Comment (gshotwell) [reply R/summarize.R]: Yup! I was following how they did it in dplyr. We could also document it in summarize, but I thought it might be more discoverable as its own documentation entry. https://github.com/tidyverse/dplyr/blob/5e9bf430745338c9abbc1ef7703754fa7e55d45f/R/manip.r#L535

**Review by jonkeane — APPROVED** [review]

**Review by nealrichardson — APPROVED** [review]

**Pull request merged**
